Implement scatter primitive for ProcessGroupNCCL

wanchaol opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #71623
* __->__ #70029
* #66745

This PR implements NCCL scatter and add scatter to ProcessGroupNCCL.

NCCL doesn’t directly provide primitives for scatter, so we need to be implemented on top of NCCL’s send/recv API.

1. In ProcessGroupNCCL.cpp, the inputTensors are first flattened, then outputTensors and inputFlattened are passed by the collective class to scatter() function in nccl.cpp.
2. In nccl.cpp, scatter is implemented using ncclSend/ncclRecv: the root rank uses a for loop to send(distribute) the inputTensors to each rank, then all the ranks receive the inputTensor from the root rank.

Differential Revision: [D33154823](https://our.internmc.facebook.com/intern/diff/D33154823/)

Commit:
Implement scatter primitive for ProcessGroupNCCL

This PR implements NCCL scatter and add scatter to ProcessGroupNCCL.

NCCL doesn’t directly provide primitives for scatter, so we need to be implemented on top of NCCL’s send/recv API.

1. In ProcessGroupNCCL.cpp, the inputTensors are first flattened, then outputTensors and inputFlattened are passed by the collective class to scatter() function in nccl.cpp.
2. In nccl.cpp, scatter is implemented using ncclSend/ncclRecv: the root rank uses a for loop to send(distribute) the inputTensors to each rank, then all the ranks receive the inputTensor from the root rank.

Differential Revision: [D33154823](https://our.internmc.facebook.com/intern/diff/D33154823/)

[ghstack-poisoned]

Commit:
Update on "Implement scatter primitive for ProcessGroupNCCL"


This PR implements NCCL scatter and add scatter to ProcessGroupNCCL.

NCCL doesn’t directly provide primitives for scatter, so we need to be implemented on top of NCCL’s send/recv API.

1. In ProcessGroupNCCL.cpp, the inputTensors are first flattened, then outputTensors and inputFlattened are passed by the collective class to scatter() function in nccl.cpp.
2. In nccl.cpp, scatter is implemented using ncclSend/ncclRecv: the root rank uses a for loop to send(distribute) the inputTensors to each rank, then all the ranks receive the inputTensor from the root rank.

Differential Revision: [D33154823](https://our.internmc.facebook.com/intern/diff/D33154823/)

[ghstack-poisoned]

Commit:
Update on "Implement scatter primitive for ProcessGroupNCCL"


This PR implements NCCL scatter and add scatter to ProcessGroupNCCL.

NCCL doesn’t directly provide primitives for scatter, so we need to be implemented on top of NCCL’s send/recv API.

1. In ProcessGroupNCCL.cpp, the inputTensors are first flattened, then outputTensors and inputFlattened are passed by the collective class to scatter() function in nccl.cpp.
2. In nccl.cpp, scatter is implemented using ncclSend/ncclRecv: the root rank uses a for loop to send(distribute) the inputTensors to each rank, then all the ranks receive the inputTensor from the root rank.

Differential Revision: [D33154823](https://our.internmc.facebook.com/intern/diff/D33154823/)

[ghstack-poisoned]

pritamdamania87 commented on torch/csrc/distributed/c10d/ProcessGroupNCCL.cpp:
```diff
@@ -2232,10 +2242,81 @@ c10::intrusive_ptr<ProcessGroup::Work> ProcessGroupNCCL::gather(
+      invalidArgument(ss.str());
+    }
+
+    inputFlattened = flatten_for_scatter_gather(inputTensors, outputTensors, size_, true);
+    check_gpu_tensors(inputFlattened);
```
It seems like we might have unnecessary copies here. Do we really need to create a flattened tensor for scatter? It seems like we anyways do `inputs[r]` in the scatter implementation, so why can't we just pass in a list of tensors to `scatter` and don't do any of the flattening or copying?

pritamdamania87 replied:
Also, doesn't this apply even to our `gather` implementation where we can avoid flattening and extra copies?

wanchaol replied:
Yeah true, I think it's a legacy issue that most of our utils function built on the top of [`collective`](https://github.com/pytorch/pytorch/blob/82c5f298edf3aa450c491fa480278d0ccdd3891e/torch/csrc/distributed/c10d/ProcessGroupNCCL.cpp#L1332), where it only takes a input tensor list and a output tensor list (the length must be the same), we utilize bunch of its checks and work initialization there, and we also do [flattening/copying in MPI impls](https://github.com/pytorch/pytorch/blob/82c5f298edf3aa450c491fa480278d0ccdd3891e/torch/csrc/distributed/c10d/ProcessGroupMPI.cpp#L632). But I do agree with you that most of these copies and flattening aren't necessary and introduce extra cost, let me see what I can do here, might need to refactor NCCL utils to avoid those copying

Commit:
Update on "Implement scatter primitive for ProcessGroupNCCL"


This PR implements NCCL scatter and add scatter to ProcessGroupNCCL.

NCCL doesn’t directly provide primitives for scatter, so we need to be implemented on top of NCCL’s send/recv API.

1. In ProcessGroupNCCL.cpp, the inputTensors are first flattened, then outputTensors and inputFlattened are passed by the collective class to scatter() function in nccl.cpp.
2. In nccl.cpp, scatter is implemented using ncclSend/ncclRecv: the root rank uses a for loop to send(distribute) the inputTensors to each rank, then all the ranks receive the inputTensor from the root rank.

Differential Revision: [D33154823](https://our.internmc.facebook.com/intern/diff/D33154823/)

[ghstack-poisoned]

Commit:
Update on "Implement scatter primitive for ProcessGroupNCCL"


This PR implements NCCL scatter and add scatter to ProcessGroupNCCL.

NCCL doesn’t directly provide primitives for scatter, so we need to be implemented on top of NCCL’s send/recv API.

1. In ProcessGroupNCCL.cpp, the inputTensors are first flattened, then outputTensors and inputFlattened are passed by the collective class to scatter() function in nccl.cpp.
2. In nccl.cpp, scatter is implemented using ncclSend/ncclRecv: the root rank uses a for loop to send(distribute) the inputTensors to each rank, then all the ranks receive the inputTensor from the root rank.

Differential Revision: [D33154823](https://our.internmc.facebook.com/intern/diff/D33154823/)

[ghstack-poisoned]

Commit:
Update on "Implement scatter primitive for ProcessGroupNCCL"


This PR implements NCCL scatter and add scatter to ProcessGroupNCCL.

NCCL doesn’t directly provide primitives for scatter, so we need to be implemented on top of NCCL’s send/recv API.

1. In ProcessGroupNCCL.cpp, the inputTensors are first flattened, then outputTensors and inputFlattened are passed by the collective class to scatter() function in nccl.cpp.
2. In nccl.cpp, scatter is implemented using ncclSend/ncclRecv: the root rank uses a for loop to send(distribute) the inputTensors to each rank, then all the ranks receive the inputTensor from the root rank.

Differential Revision: [D33154823](https://our.internmc.facebook.com/intern/diff/D33154823/)

[ghstack-poisoned]

Commit:
Update on "Implement scatter primitive for ProcessGroupNCCL"


This PR implements NCCL scatter and add scatter to ProcessGroupNCCL.

NCCL doesn’t directly provide primitives for scatter, so we need to be implemented on top of NCCL’s send/recv API.

1. In ProcessGroupNCCL.cpp, the inputTensors are first flattened, then outputTensors and inputFlattened are passed by the collective class to scatter() function in nccl.cpp.
2. In nccl.cpp, scatter is implemented using ncclSend/ncclRecv: the root rank uses a for loop to send(distribute) the inputTensors to each rank, then all the ranks receive the inputTensor from the root rank.

Differential Revision: [D33154823](https://our.internmc.facebook.com/intern/diff/D33154823/)

[ghstack-poisoned]

Commit:
Update on "Implement scatter primitive for ProcessGroupNCCL"


This PR implements NCCL scatter and add scatter to ProcessGroupNCCL.

NCCL doesn’t directly provide primitives for scatter, so we need to be implemented on top of NCCL’s send/recv API.

1. In ProcessGroupNCCL.cpp, the inputTensors are first flattened, then outputTensors and inputFlattened are passed by the collective class to scatter() function in nccl.cpp.
2. In nccl.cpp, scatter is implemented using ncclSend/ncclRecv: the root rank uses a for loop to send(distribute) the inputTensors to each rank, then all the ranks receive the inputTensor from the root rank.

Differential Revision: [D33154823](https://our.internmc.facebook.com/intern/diff/D33154823/)

[ghstack-poisoned]

pritamdamania87 commented on torch/csrc/distributed/c10d/ProcessGroupNCCL.cpp:
```diff
@@ -2261,41 +2249,38 @@ c10::intrusive_ptr<ProcessGroup::Work> ProcessGroupNCCL::scatter(
     if (inputTensors.size() != 0) {
       invalidArgument("requires empty input on non-root");
     }
+    inputTensors.emplace_back();
```
nit: Same as gather PR, probably shouldn't modify the passed in values as is.

Commit:
Update on "Implement scatter primitive for ProcessGroupNCCL"


This PR implements NCCL scatter and add scatter to ProcessGroupNCCL.

NCCL doesn’t directly provide primitives for scatter, so we need to be implemented on top of NCCL’s send/recv API.

1. In ProcessGroupNCCL.cpp, the inputTensors are first flattened, then outputTensors and inputFlattened are passed by the collective class to scatter() function in nccl.cpp.
2. In nccl.cpp, scatter is implemented using ncclSend/ncclRecv: the root rank uses a for loop to send(distribute) the inputTensors to each rank, then all the ranks receive the inputTensor from the root rank.

Differential Revision: [D33154823](https://our.internmc.facebook.com/intern/diff/D33154823/)

[ghstack-poisoned]

Commit:
Update on "Implement scatter primitive for ProcessGroupNCCL"


This PR implements NCCL scatter and add scatter to ProcessGroupNCCL.

NCCL doesn’t directly provide primitives for scatter, so we need to be implemented on top of NCCL’s send/recv API.

1. In ProcessGroupNCCL.cpp, the inputTensors are first flattened, then outputTensors and inputFlattened are passed by the collective class to scatter() function in nccl.cpp.
2. In nccl.cpp, scatter is implemented using ncclSend/ncclRecv: the root rank uses a for loop to send(distribute) the inputTensors to each rank, then all the ranks receive the inputTensor from the root rank.

Differential Revision: [D33154823](https://our.internmc.facebook.com/intern/diff/D33154823/)

[ghstack-poisoned]

Commit:
Update on "Implement scatter primitive for ProcessGroupNCCL"


This PR implements NCCL scatter and add scatter to ProcessGroupNCCL.

NCCL doesn’t directly provide primitives for scatter, so we need to be implemented on top of NCCL’s send/recv API.

1. In ProcessGroupNCCL.cpp, the inputTensors are first flattened, then outputTensors and inputFlattened are passed by the collective class to scatter() function in nccl.cpp.
2. In nccl.cpp, scatter is implemented using ncclSend/ncclRecv: the root rank uses a for loop to send(distribute) the inputTensors to each rank, then all the ranks receive the inputTensor from the root rank.

Differential Revision: [D33154823](https://our.internmc.facebook.com/intern/diff/D33154823/)

[ghstack-poisoned]
